Core/clim-basic/utils.lisp

```diff
@@ -808,161 +808,76 @@ a flag CLOSED is T then beginning and end of the list are consecutive too."
-                               (setf best-breaks (list* current-break breaks)
-                                     best-cost (+ current-cost remaining-cost))))))
-                        ((null best-cost)
-                         (return-from breaks-rec
-                           (if (> offset 0)
-                               (multiple-value-bind (breaks remaining-cost)
-                                   (breaks-rec 0 start end i)
-                                 (return-from breaks-rec
-                                   (values (list* start breaks)
-                                           (+ remaining-cost current-margin))))
-                               (let* ((char-breaks (%line-breaks-1 string width
-                                                                   offset margin
-                                                                   start opportunity))
-                                      (new-start (alexandria:last-elt char-breaks)))
-                                 (multiple-value-bind (breaks remaining-cost)
-                                     (breaks-rec 0 new-start end i)
-                                   (values (append char-breaks breaks) remaining-cost))))))
-                        (t #1=(return-from breaks-rec
-                                (values best-breaks best-cost))))
-                finally #1#)))
-    (breaks-rec initial-offset start end 0)))
-
-;;; Implementing line breaking as defined in Unicode[1] is left as an
-;;; excercise for the reader. When implemented it should be wired to
-;;; the strategy designated by T. Current approach break on space, but
-;;; in "real" languages lines may /need to/ or /must to/ be split under
-;;; various conditions - hyphen, hard space, infix numeric separators
-;;; etc. Optimally this should be implemented in cl-unicode and used
-;;; from there.  [1] https://unicode.org/reports/tr14/ -- jd 2019-01-08
-(defun line-breaks (string width &key (break-strategy t) initial-offset margin (start 0) end)
-  "Function takes a string and returns a list of indexes where it should be split.
+;;; Returns an index of the character _before_ which we should break
+;;; the line (may be used as END argument in SUBSEQ). When FROM-END is
+;;; T, then we return the rightmost opportunity which does not violate
+;;; hard line breaks. Second value indicates the kind of a line break.
+(defun line-end (string break-fn start limit from-end)
+  (loop with opportunity = nil
+        for i from start below limit
+        do (case (maybe-funcall break-fn string start i)
+             (:hard (return (values (1+ i) :hard)))
+             (:soft (if from-end
+                        (setf opportunity (1+ i))
+                        (return (values (1+ i) :soft)))))
+        finally
+           (if (null opportunity)
+               (return (values limit :emergency))
+               (return (values opportunity :soft)))))
+
+(defun line-breaks (string width
```

Comment:
Maybe it didn't load everything properly, but `(line-breaks "foo bar baz" 8)` seems to go into an infinite loop.